### PR TITLE
CHE-7162: Improve branch search mechanism in Git checkout command

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
Improve branch search mechanism in Git checkout command
Fixes bug when wrong Git upstream can be set when importing a project with specified branch.
### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/7162

Improve branch search mechanism in Git checkout command

#### Release Notes

N/A

#### Docs PR

N/A

### CODE REVIEW

**Comment by vinokurig** [comment]
ci-test

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 1]
```diff
@@ -379,27 +382,51 @@ public void checkout(CheckoutParams params) throws GitException {
...
-              branchList(LIST_REMOTE)
+              branches
                   .stream()
-                  .filter(branch -> branch.getName().contains(name))
+                  .filter(
+                      branch ->
+                          remotes
+                              .stream()
+                              .anyMatch(
+                                  remote -> branch.getName().endsWith(remote + "/" + branchName)))
```
Comment (tolusha) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: I guess there is a way to use the only stream iteration.

Comment (vinokurig) [reply wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: done

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 2]
```diff
@@ -379,27 +382,49 @@ public void checkout(CheckoutParams params) throws GitException {
...
+                                      ("refs/remotes/" + remote + "/" + branchName)
+                                          .equals(branch.getName())))
+                  .collect(Collectors.toList());
+          if (remotes.size() > 1 || remotes.size() == 0) {
+            throw new GitException(String.format(ERROR_CHECKOUT_BRANCH_NAME_NOT_FOUND, branchName));
+          }
+
           Optional<Branch> remoteBranch =
-              branchList(LIST_REMOTE)
+              branches
```
Comment (tolusha) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: Accordingly to the all entries in the `branches` are equal to the `branch.getName()`
```
("refs/remotes/" + remote + "/" + branchName)
 +                                          .equals(branch.getName())
```

**Review by tolusha — APPROVED** [review]

**Comment by tolusha** [comment]
ci-test

**Comment by tolusha** [comment]
ci-test

**Comment by tolusha** [comment]
ci-test

**Comment by skabashnyuk** [comment]
how does this improvement work?

**Comment by vinokurig** [comment]
When checkout request received with `name` parameter, and there is no local branch that matches `name` parameter it begins to search it in remote branches. If the branch is present in several remotes it throws exception (native Git does in the same way).
As for now branches are filtered by simple `contains` condition so it may cause a problem when wrong branch can be set to upstream

**Comment by vinokurig** [comment]
ci-test

**Comment by vinokurig** [comment]
ci-test

**Comment by vinokurig** [comment]
All Git selenium test passed successfully after rerunning localy

**Review by vparfonov — APPROVED** [review]

**Review by artaleks9 — APPROVED** [review]

**Pull request merged**